main/src/main/scala/sbt/internals/parser/SbtParser.scala

```diff
@@ -6,14 +6,100 @@ import java.io.File
 
 import sbt.internals.parser.SbtParser._
 
-import scala.reflect.runtime.universe._
+import scala.compat.Platform.EOL
+import scala.reflect.internal.util.BatchSourceFile
+import scala.reflect.io.VirtualDirectory
+import scala.tools.nsc.interactive.RangePositions
+import scala.tools.nsc.{ CompilerCommand, Global }
+import scala.tools.nsc.reporters.StoreReporter
 
 private[sbt] object SbtParser {
   val END_OF_LINE_CHAR = '\n'
   val END_OF_LINE = String.valueOf(END_OF_LINE_CHAR)
   private[parser] val NOT_FOUND_INDEX = -1
   private[sbt] val FAKE_FILE = new File("fake")
   private[parser] val XML_ERROR = "';' expected but 'val' found."
+
+  private val XmlErrorMessage =
+    """Probably problem with parsing xml group, please add parens or semicolons:
+      |Replace:
+      |val xmlGroup = <a/><b/>
+      |with:
+      |val xmlGroup = (<a/><b/>)
+      |or
+      |val xmlGroup = <a/><b/>;
+    """.stripMargin
+
+  private final val defaultClasspath =
+    Path.makeString(IO.classLocationFile[Product] :: Nil)
+
+  /**
+   * Provides the previous error reporting functionality in
+   * [[scala.tools.reflect.ToolBox]].
+   *
+   * This is a sign that this whole parser should be rewritten.
+   * There are exceptions everywhere and the logic to work around
+   * the scalac parser bug heavily relies on them and it's tied
+   * to the test suite. Ideally, we only want to throw exceptions
+   * when we know for a fact that the user-provided snippet doesn't
+   * parse.
+   */
+  private[sbt] class ParserStoreReporter extends StoreReporter {
+    def throwParserErrorsIfAny(fileName: String): Unit = {
+      if (parserReporter.hasErrors) {
+        val seq = parserReporter.infos.map { info =>
+          s"""[$fileName]:${info.pos.line}: ${info.msg}"""
+        }
+        val errorMessage = seq.mkString(EOL)
+        val error: String =
+          if (errorMessage.contains(XML_ERROR))
+            s"$errorMessage\n${SbtParser.XmlErrorMessage}"
+          else errorMessage
+        throw new MessageOnlyException(error)
+      }
+    }
+  }
+
+  private[sbt] final val parserReporter = new ParserStoreReporter
+
+  private[sbt] final lazy val defaultGlobalForParser = {
+    import scala.reflect.internal.util.NoPosition
+    val options = "-cp" :: s"$defaultClasspath" :: "-Yrangepos" :: Nil
+    val reportError = (msg: String) => parserReporter.error(NoPosition, msg)
+    val command = new CompilerCommand(options, reportError)
+    val settings = command.settings
+    settings.outputDirs.setSingleOutput(new VirtualDirectory("(memory)", None))
+
+    // Mixing RangePositions is necessary, otherwise global ignores -Yrangepos
+    val global = new Global(settings, parserReporter) with RangePositions
```

Comment:
Isn't the purpose of the toolbox to provide some abstraction instead of strongly coupling the implementation of sbt with Scala compiler? It is possible to reuse the toolbox, or would that incur initialization penalty?

/cc @adriaanm, @retronym